src/main/java/com/openlattice/authorization/EdmAuthorizationHelper.java

```diff
@@ -112,15 +112,99 @@ public EdmAuthorizationHelper( EdmManager edm, AuthorizationManager authz ) {
+            EnumSet<Permission> requiredPermissions,
+            Set<Principal> principals ) {
+        if ( linkingEntitySet.getLinkedEntitySets().isEmpty() ) {
+            return Maps.newHashMap();
+        } else {
+            final var authorizedPropertyTypesOfNormalEntitySets = getAuthorizedPropertiesOnEntitySets(
+                    linkingEntitySet.getLinkedEntitySets(), requiredPermissions, principals );
+
+            final var authorizedProperties = authorizedPropertyTypesOfNormalEntitySets
+                    .values().stream()
+                    .reduce( ( properties, nextProperties ) ->
+                            Maps.difference( properties, nextProperties ).entriesInCommon() )
+                    .orElse( Maps.newHashMap() );
+
+            return linkingEntitySet.getLinkedEntitySets().stream().collect( Collectors.toMap(
+                    Function.identity(),
+                    esId -> authorizedProperties
+            ) );
+        }
+    }
+
+    /**
+     * Note: entitysets are assumed to have same entity type
+     */
+    public Map<UUID, Map<UUID, PropertyType>> getAuthorizedPropertyTypesByNormalEntitySet(
+            EntitySet linkingEntitySet,
+            Set<UUID> selectedProperties,
+            EnumSet<Permission> requiredPermissions ) {
+        if ( linkingEntitySet.getLinkedEntitySets().isEmpty() ) {
+            return Maps.newHashMap();
+        } else {
+            final var authorizedPropertyTypesOfNormalEntitySets = getAuthorizedPropertyTypes(
+                    linkingEntitySet.getLinkedEntitySets(),
+                    selectedProperties,
+                    requiredPermissions,
+                    Principals.getCurrentPrincipals() );
+
+            final var authorizedProperties = authorizedPropertyTypesOfNormalEntitySets
+                    .values().stream()
+                    .reduce( ( properties, nextProperties ) ->
+                            Maps.difference( properties, nextProperties ).entriesInCommon() )
+                    .orElse( Maps.newHashMap() );
+
+            return linkingEntitySet.getLinkedEntitySets().stream().collect( Collectors.toMap(
+                    Function.identity(),
+                    esId -> authorizedProperties
+            ) );
+        }
+    }
+
+    public Map<UUID, PropertyType> getAuthorizedPropertyTypesOfLinkingEntitySet(
+            EntitySet linkingEntitySet,
+            EnumSet<Permission> requiredPermissions ) {
+        return getAuthorizedPropertyTypesOfLinkingEntitySet(
+                linkingEntitySet, requiredPermissions, Principals.getCurrentPrincipals() );
+    }
+
+    public Map<UUID, PropertyType> getAuthorizedPropertyTypesOfLinkingEntitySet(
+            EntitySet linkingEntitySet,
+            EnumSet<Permission> requiredPermissions,
+            Set<Principal> principals ) {
+        if ( linkingEntitySet.getLinkedEntitySets().isEmpty() ) {
+            return Maps.newHashMap();
+        } else {
```

Comment:
this doesn't need to be in an else

Reply:
Yeah, I would change to isNotEmpty() and then just have a default return as the cleaner approach.

Reply:
Ah, yes, thanks!